src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+or [abstract members][].
+
+### Implements
+
+You can optionally use the `implements` clause to introduce
+a subtype relationship on an extension type, making *all* members
+of the representation type available to the extension type:
+
+```dart
+extension type NumberI(int i) 
+  implements int{
+  // ...
+}
+```
+
+An extension type can only implement its representation type
+or a supertype of its representation type.
+
+This is technically NOT an inheritance relationship
+like the one `implements` introduces for classes,
+but rather an [applicability][] relationship like that between an
+extension method and its `on` type.
+Members that apply to the supertype are applicable to the subtype,
+unless the supertype has a more specific definition for the member.
+
+#### `@redeclare`
+
+Replacing a method that shares a name with a member of the supertype
+is not an override relationship for extension types, but a *redeclaration*.
+Use the `@redeclare` annotation to tell the compiler you are *knowingly* choosing
+to use the same name as a supertype's member definition, not just accidentally
+using the same name:
+
+```dart
+class C {
+  void m() {}
+}
+
+extension type E(C c) implements C {
+  @redeclare
+  void m() {}
+}
+```
+
+If you don't use the annotation, the compiler will choose the [more specific][]
+definition wherever the name is invoked, which might not be the behavior you expect.
+
+## Usage
+
+To use an extension type, create an instance the same as you would with a class:
+
+```dart
+extension type NumberE(int value) {
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+
+  NumberE get i => value;
+  bool isValid() => !value.isNegative;
+}
+
+void testE() { 
+  var num = NumberE(1);
+}
+```
+
+Then you can invoke members on the object as you would with a class object.
+Any attempt to treat the extension type object as an object
+of the representation type, or vice versa, will result in compile time errors
+(unless that behavior is explicitly defined for the extension type,
+or it is [transparent](#transparency)).
+For example:
+
+```dart
+void testE() { 
+  var num = NumberE(1);
+  int num2 = NumberE(2); // Error: Can't assign 'NumberE' to 'int'.
+  
+  num.isValid(); // Ok: Extension member invocation.
+  num.isNegative(); // Error: 'NumberE' does not define 'int' member 'isNegative'.
+  
+  var sum = num + num; // Ok: 'NumberE' defines '+'.
+  var diff = num - num; // Error: 'NumberE' does not define 'int' member '-'.
+  var diff2 = num.value - 2; // Ok: Can access representation object with reference.
+  var sum2 = num + 2; // Error: Can't assign 'int' to parameter type 'NumberE'. 
+  
+  List<NumberE> numbers = [
+    NumberE(1), 
+    num.i, // Ok: 'i' getter returns type 'NumberE'.
+    1, // Error: Can't assign 'int' element to list type 'NumberE'.
+    1 as NumberE, // Ok: Dynamic type cast recognizes representation type.
+  ];
+}
+```
+
+#### Transparency
```

Comment:
If I'm understanding correctly (from this and your other comments on this section) would it be better to say:

"An extension type that implements its representation type, or a supertype of its representation type, can be _thought of_ as transparent, because it provides a kind of overarching access to underlying type members (at least _somewhere_ in the hierarchy / either those of the direct representation type or somewhere higher in the type hierarchy) that is otherwise _not_ the default for the intrinsic isolated nature of extension type interfaces"?

Not that that's exactly what I'll write, but is my understanding correct?

Reply:
> It's also possible to have extension types that implement a _supertype_ of their representation type.

True. I did not push on emphasizing this point because I think it's a natural refinement. The crucial message to communicate is that there are two equally valid, but very different, ways to use an extension type: It can be used (1) to extend the interface of an existing type (adding extra members), and (2) to create an unrelated type with an unrelated interface.

An example of the former is `FancyString` that adds extra methods to `String`; we surely want to be able to use an expression of this type in all the ways that `String` supports. An example of the latter is `IdNumber` that provides an interface which is suitable for ID numbers, but allows us to use plain (and cheap) `int` objects for that purpose; in this case we definitely do not want to silently include the interface of `int` as part of the interface of `IdNumber`.

I think those two modes of operation are so different that they should be established _conceptually_, thus clearly communicating that each of them is OK, but each of them comes with its own conceptualization and software engineering rules.

It would be helpful to support this conceptualization by means of a well-chosen terminology, and 'transparent' seems fine to me. So `FancyString` is transparent because we "can see" the underlying type `String`, and `IdNumber` is 'not transparent' because it is an important part of being an ID number that it does not do all the `int` things.

On top of that, we can refine: It's possible for an extension type `E` to have a representation type `R1` and a clause `implements ... R2 ...` where `R1` is a proper subtype of `R2`, e.g., because `R1` is private and `R2` defines the part of the interface of `R1` that we wish to provide to clients. We can also have a mostly-transparent extension type that adds some new members and _adapts_ others by redeclaring the given name (e.g., in order to use stricter types on some parameters of a method, or different default values, etc). Another way to be mostly transparent is to implement a type which is a proper supertype of the representation type.

My take on the refinements was that there may or may not be room for that level of detail here. If we can help developers to firmly establish the notion of an extension type as transparent or not transparent then it won't be a problem to deal with everything in between.

Reply:
Ok, I think I figured out how to handle this (commit coming up!).

1. I'll include Lasse's points about the different potential `implements` scenarios with examples in the "Implements" section. Since that section comes first, this will hopefully set the precedent that we are *not* implying implementing the representation type or nothing at all are your only options. 

2. Then later on in the "Usage" section, I'll describe Erik's crucial use cases, transparent and non-transparent. I'll include information about the other `implements` situations in the "transparent" case section to not gatekeep the scenario too much. I'll generally try not to be too strict about "transparency" as a title, but more just "you can _think_ of these as transparent because..." (also based on Erik's comment here). 

Thanks for the attention to detail!